src/scoperesolver.js

```diff
@@ -237,34 +230,156 @@ class ScopeResolver extends events.EventEmitter {
+      let seen = [];
+      let visit = roleId => {
+        if (done.has(start)) {
+          return;
+        }
+        if (seen.indexOf(roleId) !== -1) {
+          return seen.slice(seen.indexOf(roleId)).concat([roleId]);
+        }
+        seen.push(roleId);
+        for (let edge of edges[roleId] || []) {
+          let cycle = visit(edge);
+          if (cycle) {
+            return cycle;
+          }
+        }
+        done.add(seen.pop());
+      };
+
+      let cycle = visit(start);
+      if (cycle) {
+        throw new Error(`Found cycle in roles: ${cycle.map(c => `${c}*`).join(' -> ')}`);
+      }
+    });
+  };
+
+  /**
+   * Build a resolver which, given a set of scopes, will return the expanded
+   * set of scopes based on the given roles.  Roles are an array of elements
+   * {roleId, scopes}.
+   */
+  buildResolver(roles) {
+    ScopeResolver.cycleCheck(roles);
+
+    // encode the roles as rules, including the `assume:` prefix, and marking up
+    // the expansions of any parameterized scopes as {scope, index}, where index
+    // is the index in the input at which the replacement begins (the index of
+    // the `*`)
+
+    let rules = [];
+    roles.forEach(({roleId, scopes}) => {
+      let pattern = `assume:${roleId}`;
+      if (pattern.endsWith('*')) {
+        const index = pattern.length - 1;
+        rules.push({pattern, scopes: scopes.map(scope =>
+          PARAMETER.test(scope) ?  {scope, index} : scope)});
+      } else {
+        rules.push({pattern, scopes});
+      }
+    });
+
+    let dfa = generateTrie(rules);
+
+    return (inputs) => {
+      let scopes = inputs;
+      let seen = new Set();
+      let i = 0;
+      while (i < scopes.length) {
+        let scope = scopes[i++];
+        // if this scope is not going to expand, ignore it
+        if (!ASSUME_PREFIX.test(scope)) {
+          continue;
+        }
+        // if we have seen this scope, ignore it
+        if (seen.has(scope)) {
+          continue;
+        }
+        seen.add(scope);
+        // eecute the DFA and expand any parameterizations in the result, then add
+        // the newly expanded scopes to the list of scopes to expand (recursively)
+        scopes = scopes.concat(
+          executeTrie(dfa, scope).map(expansion => {
+            if (typeof expansion === 'string') {
+              return expansion;
+            }
+            const {scope: expandedScope, index} = expansion;
+
+            if (scope.length < index) {
+              return expandedScope.replace(PARAMETER_TO_END, '*');
+            } else if (scope.endsWith('*')) {
+              return expandedScope.replace(PARAMETER_TO_END, scope.slice(index));
+            } else {
+              return expandedScope.replace(PARAMETER_G, scope.slice(index));
+            }
+          }));
+      }
+
+      // TODO: normalize should remove equivalent scopes too
+      // https://github.com/taskcluster/taskcluster-lib-scopes/pull/17
+      scopes = [...new Set(scopes)];
```

Comment:
If instead of using a `Set` for `seen` you used a sorted list from `normalizeScopeSet` you could do a binary search to see if it contains any scope that is greater than the given scope. In fact it would be fairly cheap (similar to a merge operation) to find the scopes from `executeTrie(dfa, scope).map(...` that are not covered by another scope in `seen` and add those to `scopes`.

Then you would instantiate `seen = normalizeScopeSet(input.sort(scopeCompare))` and `scopes = [...seen]`, obviously skipping the `if (seen.has(scope)) { continue`.

This would fix the case where `assume:*` takes a look time to expand, because `assume:*` covers all other `assume:<something>` scopes, hence, they wouldn't need to be added to `scopes`.